### ISSUE TITLE
All tags and notes duplicated

### ISSUE THREAD

**Issue by boxoffice** [standardnotes/forum#166]
Sometimes my tags and note duplicate them-selfs. So today, I put in the effort to tidy up my notes and deleted all doublets one by one using the web app. After the work was done I signed out. I signed back in. Now all tags and notes had duplicated :(  This is so frustrating.
Is there any way to remove these doublets in a convenient way and to prevent this type of random duplication in the future? 

Thanks for your help in advance.

**Comment by mobitar** [comment]
Hey, sorry about that. I have it on my todo to work on a "duplicate manager", given that duplication will remain a familiar friend due to encrypted sync. I've bumped it up on my queue. Perhaps we'll have something soon :)

**Comment by boxoffice** [comment]
Hi mobitar, thanks for the fast response. What does "soon" mean, March, before summer, in 2018 or before 2020? 
In the mean time, is there a way of using standard notes that will avoid duplication from happening? As it is now, standard notes is unusable for me. Which is very sad because I really like it a lot.

**Comment by mobitar** [comment]
You can avoid duplication by making sure the devices you use SN on have a healthy connection. If they don't, and you make changes on one, and changes on another, you will see duplication.

I personally never experience duplication, but I am always on a healthy connection. If you find a way to replicate the behavior consistently, let me know.

Soon in this case means as soon as I can. It's a high priority.

**Comment by jaco-l** [comment]
I had this issue yesterday too. It's extremely annoying indeed. I noticed that not all notes duplicated themselves, but about 70% of them. Also the duplicated notes had a double tag like #Todo #Todo and the tag Todo was now mentioned twice in the left column.  I did not observe exactly what triggered the duplication, but some red colored warnings came up in that left column. I can also not replicate the incident, but I hope this adds some information nevertheless.

**Comment by mobitar** [comment]
Ok, the Batch Manager/Duplicate Deleter extension is now live on the web app. Simply click Extensions, then find Batch Manager, and press Open.

**Comment by cookies8710** [comment]
Works great on the Notes, but still leaves Tags with 0 occurrences.

**Comment by mobitar** [comment]
Strange. Might be a references issue. That is, essentially, if one tag has 5 notes and the duplicate has 4, they might not be marked duplicates.

**Comment by Ossur** [comment]
Just ran into a duplication issue today (first time in my 6 months of using Standard notes). Here's what happened: 
1. In the browser client I added a new tag and two new notes. Also did archive several other notes with other tags
2. Went outside and wanted to check the new notes in the Android client but the new tag didn't show up there, probably because the phone was in the process of loosing its WiFi connection when I opened the client.  
3. Went back inside, restarted the Android client, and could now see the new tag. However, all the old tags are now duplicated where the archived notes are in one version but are archived in the other. It's as if both old and new versions are showing up simultaneously on my phone.
4. Tried restarting and re-logging into both clients and then the duplicates also showed up in the browser client.
5. Unfortunately the "Find duplicates" button in the Batch Manager didn't work for my case (all entries disappear when I clicked it). But it was helpful in enabling me to quickly delete the duplicates. However, it would be very helpful to be able to order the items by name, so you could be really sure you're not deleting both versions by accident.

**Comment by mobitar** [comment]
Thanks for the heads up. So in this case the duplication wasn't of all your data, but just pieces of your data that you recently changed? I do believe in your circumstances, with the network transitioning state, duplication is to be expected, as this occurs when the client receives two sets of changes that are in conflict to one another. In which case, the data is duplicated. 

Can you elaborate what you mean by "all entries disappeared when I clicked it" in the Batch Manager?

**Comment by Ossur** [comment]
No, all tags and notes were duplicated except the ones I created in step 1. (I just realized I haven't yet deleted the duplicate notes - they just have no tags now, since I deleted all the duplicate tags.)

In the Batch Manager, before I click the "Find duplicates" button, I can see all the notes in a table. After I press the button, the whole table (along with the headers) disappears, the "Select all" button disappears too, but the "Find duplicates" is replaced by "Hide duplicates" button. When I click "Hide duplicates" the table reappears. 

Should duplication really be expected if I didn't do any edits on the Android client? (Unless I did something small change accidentally without noticing)

**Comment by jaco-l** [comment]
I have not experienced this issue again since my above report (Feb 28) and I wanted to offer a possible hint to the cause. This duplication error seems more likely to occur when you clients connect to the internet at different and topologically far removed points. In my case, I used VPN's connecting to different countries from my laptops. In Ossur's case, Wifi/home network vs. telco's mobile data network.  
Could it be that AWS serves such distant, simultaneous client connections from two different data centres by quickly replicating the database with notes and keeping them in sync? Maybe this type of syncing is sometimes error prone or too slow? 
I am in no way an expert in this area, but only wanted to put this on your radar. I noticed that my duplication issues stopped since pointing my laptop VPN's to the same VPN server location. It could also be coincidence.

**Comment by mobitar** [comment]
@Ossur regarding the disappearing table in Batch Manager, it might just be that no duplicates were found. (Duplicate checking also takes into consideration the relationships an item has, so a note for example would not be considered a duplicate if they had the same text but different relationships with tags).

@jaco-l yeah, it's definitely to do with wavering network conditions. It's just so hard to reproduce these issues, or even reasoning about them through thought experiments. I've tested using simulated poor network conditions, and don't encounter any issues.

**Comment by effieeee** [comment]
Closing as it's been quite a long while since this issue, the batch manager has also been replaced by the multiple selection capability and context menu. Please reopen if you still encounter other issues with duplication! 🙏
